### ISSUE TITLE
cr_log_* produces the error "Invalid byte sequence" for legal character codes on Windows

### ISSUE THREAD

**Issue by MIvanchev** [Snaipe/Criterion#216]
Executing the block

```
wchar_t str[] = { 0x0105, L'\0' };
cr_log_error("This output is going to crash %ls", str);
```

produces the following error on Windows 7 x64

`[ERR ] Could not write the "message" message down the event pipe: Illegal byte sequence.`

0x0105 is the code of the character "ą" (Latin Small Letter A with ogonek).

**Comment by Snaipe** [comment]
I can also reproduce this on Linux, so I'm guessing either something is wrong with the `%ls`, or the utf8 conversion fails somewhere.

**Comment by am11** [comment]
I attached debugger and it seems like error is coming from upstream dependency https://github.com/nanopb/nanopb/blob/6a6903b/pb_encode.c#L336 (probably when it hits `\0`).

@Snaipe, I even get access violation with:

```c++
#include "./includes/criterion/criterion.h"

int main() {
  cr_log_error("This output is going to crash with access violation\0");
  return 0;
}
// compiled with cl.exe using:
//
// cl test.cpp /EHsc /link ./build/Release/criterion.lib
```

thrown upon acquiring the mutex lock (`EnterCrititcalSection`) at: https://github.com/nanomsg/nanomsg/tree/7e12a20/src/utils/mutex.c#L44-L46.

**Comment by Snaipe** [comment]
No -- that's not quite it: cr_log_error can only be run from a test worker (i.e. not main()). The segfault occurs because nothing has been initialized.

It seems the issue is twofold:

1. locale-aware functions are failing in test workers. This is because since we switched to BoxFort, the locale isn't initialized by setlocale(LC_ALL, "") as the runner does, because we execve() our workers. So, we need to call that again in the worker main.
2. cr_asprintf was incorrectly returning -errno rather than -1 since it was reimplemented, which made cr_log think the call succeeded.

I've pushed a `fix/216` branch. @MIvanchev can you test if it works?

**Comment by am11** [comment]
> not main()

Ops, forgot about build-in entry point. 😀

I tested with fix/216 on win10 and it seems to not crashing. However, the stdout is printing `a` instead of `ą`:

```cmd
[ERR ] This output is going to crash a
[====] Synthesis: Tested: 1 | Passing: 1 | Failing: 0 | Crashing: 0 
```

We would need to set the code page for workers and main (using `SetConsoleOutputCP` or something more platform agnostic?).

I have confirmed with the following patch, on top of your branch, that setting code page to [65001](https://msdn.microsoft.com/library/windows/desktop/dd317756?f=255&MSPPError=-2147217396) prints the cyrillic characters correctly in console:

<details> 
  <summary>Click to expand</summary> 

```diff
diff --git a/src/core/runner_coroutine.c b/src/core/runner_coroutine.c
index 2ad597b..a9ea3a69 100644
--- a/src/core/runner_coroutine.c
+++ b/src/core/runner_coroutine.c
@@ -183,6 +183,10 @@ err:
     return rc;
 }
 
+#if WIN32
+#include <windows.h>
+#endif
+
 static int run_test_child(void)
 {
 #ifndef ENABLE_VALGRIND_ERRORS
@@ -192,6 +196,10 @@ static int run_test_child(void)
     /* the child also needs to initialize the locale, otherwise
        wchat_t -> char conversions will fail. */
     setlocale(LC_ALL, "");
+#ifndef WIN32
+#else
+    SetConsoleOutputCP(65001);
+#endif
 
     struct criterion_test test;
     memset(&test, 0, sizeof (test));
diff --git a/src/entry/params.c b/src/entry/params.c
index 04fd47b..455f2b2 100644
--- a/src/entry/params.c
+++ b/src/entry/params.c
@@ -37,6 +37,10 @@
 #include "common.h"
 #include "err.h"
 
+#if WIN32
+#include <windows.h>
+#endif
+
 #if ENABLE_NLS
 # include <libintl.h>
 #endif
@@ -279,7 +283,11 @@ CR_API int criterion_handle_args(int argc, char *argv[],
         { 0,                 0,                 0, 0   }
     };
 
+#ifndef WIN32
     setlocale(LC_ALL, "");
+#else
+    SetConsoleOutputCP(65001);
+#endif
 #if ENABLE_NLS
     textdomain(PACKAGE "-test");
 #endif
```
</details>

**Comment by Snaipe** [comment]
> We would need to set the code page for workers and main (using SetConsoleOutputCP or something more platform agnostic?).

Good point. I think only the runner needs to call that though, as it is responsible for writing to the console, and both it and the workers share the same console object.

**Comment by Snaipe** [comment]
(it seems that SetConsoleOutputCP not being called is what the [appveyor failure](https://ci.appveyor.com/project/Snaipe/criterion/build/2.3.2_b1753-fix/216/job/ef388q0vo7va9v7i#L1528) is about)

**Comment by am11** [comment]
> only the runner needs to call that

This is true, though in `params.c`, I had to at least delete / ifdef out the `setlocale(LC_ALL, "");` line to get the UTF-8 output.

<ins>Considerations for consumer:</ins>

Even with this fix, consumer would need to:

* compile *their* project with `/utf-8` flag if they have UTF-8 string on their side. For example in `cr_log_error("Format ą in string: %s", str);`, the part `Format ą in string` will be printed as `Format . in string` without this switch. Alternatively, they can use `#pragma execution_character_set("utf-8")` in *their* code. Adding these in Criterion projects has no effect from consumer side (as it is always overridden by the executable project).
* consider that Criterion exposes multi-byte UTF-8 `char*` APIs, and not wide character `wchar*` based APIs. Windows console APIs are not multibyte UTF-8 based, the natural fit is UTF-16 wchar. However, to satisfy the requirement of Criterion (or generally any UTF-8 only lib), consumer would need to do this conversion before calling Criterion API:

    ```c
    #include <criterion.h>
    #include <Windows.h>

    Test(_,__) {
        wchar_t wstr[] = { 0x0105, L'\0' };

        // conversion to multibyte (Windows-only)
        int size_needed = WideCharToMultiByte(CP_UTF8, 0, wstr, wcslen(wstr), NULL, 0, NULL, NULL);
        char *str = (char*)malloc((size_needed + 1) * sizeof(char));
        WideCharToMultiByte(CP_UTF8, 0, wstr, wcslen(wstr), str, size_needed, NULL, NULL);
        str[size_needed] = '\0';

        cr_log_error("This will print ą before and after colon: %s", str); // note %s, not %ls
    }
    ```

If we expose wide character based APIs in criterion, which will eventually make the macro call different overload with `const wchar_t*` argument (as opposed to `const char*`) and use `fwprintf` (as opposed to `fprintf`) to print the result, then this conversion won't be necessary in user land and it will work cross platform.

---

We did similar thing in libsass https://github.com/sass/libsass/commit/8dbcfd2d88d5afdcd1cba7129578c214ffbb1297, where libsass needed to get the CWD from OS via CPP's std::wstring using the `wgetcwd` API. libSass itself is UTF8-only and demands consumers to provide multibyte strings (like Criterion and many libs). However, after this exercise, we did realize that it would improve the performance on Windows if we can avoid conversion overhead by exposing more natural 'W' variant of our APIs, following the STL and other standard libraries.

**Comment by Snaipe** [comment]
I would have thought that `%ls` would have done the right thing here.

I think at this point we're having too many issues with the platform differences of the printf function family and might as well use/fork [trio](https://daniel.haxx.se/projects/trio/) for this purpose.

Do you have any idea why `setlocale(LC_ALL, "");` wouldn't be working here? I would expect the locale needs to be at least initialized before setting the console codepage.

**Comment by am11** [comment]
It is painful in a way that Windows console has separate APIs for wide characters; WriteConsoleW etc.

I think instead of `SetConsoleOutputCP`, using [`_setmode`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setmode) (file translation mode) on `stderr` for Unicode is more appropriate. 

This example demonstrate the caveats:

```c
// test.c
//
// cl test.c /utf-8

#include <wchar.h>
#include <stdio.h>
#include <locale.h>
#include <io.h> //for _setmode
#include <fcntl.h> //for _O_U16TEXT

int main(void)
{
   setlocale(LC_ALL, "");
   setlocale(LC_CTYPE, "en_US.UTF-8");


   wchar_t str[] = { 0x0105, L'\0' };

   int oldmode = _setmode(_fileno(stderr), _O_U16TEXT);
   fwprintf(stderr, L"This will print accented ą before and after colon: %ls\r\n", str);

   _setmode(_fileno(stderr), oldmode);
   fprintf(stderr, "This will print PLAIN a before and after colon: %ls\r\n", str);

   // but if I use UTF-8 mode, it will hang the console forever. There is no gurantee that
   // the multi-byte chars work for all Unicode characters in U8 mode
   // _setmode(_fileno(stderr), _O_U8TEXT);
   // fprintf(stderr, "This will neither print accented ą nor PLAIN a before and after colon: %ls\r\n", str);

   return 0;
}
```

IMO, it will make sense to provide wchar_t overloads for example `void criterion_log_noformat(enum criterion_severity severity, const wchar_t *msg)`, to avoid explicit conversion to multibyte UTF-8.

```c
// criterion/src/log/logging.c

#ifdef _Win32
#include <io.h> //for _setmode
#include <fcntl.h> //for _O_U16TEXT
#endif

// the `L` variant of formats
# define LLOG_FORMAT      L"[%s%s%s] %s"
# define LERROR_FORMAT    L"[%s%s%s] %s%s%s%s"

// the `w` variant of log functions
void criterion_log_noformat(enum criterion_severity severity, const wchar_t *msg)
{
  ...

#ifdef _Win32
    int oldmode = _setmode(_fileno(stderr), _O_U16TEXT);
#endif

    if (severity == CR_LOG_ERROR) {
        fwprintf(stderr, _(LERROR_FORMAT),
  ...
    } else {
        fwprintf(stderr, _(LLOG_FORMAT),
  ...
#ifdef _Win32
  _setmode(_fileno(stderr), oldmode);
#endif
}
```

**Comment by am11** [comment]
> Do you have any idea why setlocale(LC_ALL, ""); wouldn't be working here? I would expect the locale needs to be at least initialized before setting the console codepage.

Certainly, the only place we need is runner like you mentioned. The one in params.c doesn't seem to be needed. Here is updated patch:

```diff
~\Source\Repos\Criterion [fix/216 ≡ +0 ~4 -0 !]> git diff *.c
diff --git a/src/core/runner_coroutine.c b/src/core/runner_coroutine.c
index 2ad597b..e879b69 100644
--- a/src/core/runner_coroutine.c
+++ b/src/core/runner_coroutine.c
@@ -52,6 +52,10 @@
 # define RUNNING_ON_VALGRIND    0
 #endif

+#ifdef _WIN32
+#include <windows.h>
+#endif
+
 typedef const char *const msg_t;

 #ifdef ENABLE_NLS
@@ -193,6 +197,10 @@ static int run_test_child(void)
        wchat_t -> char conversions will fail. */
     setlocale(LC_ALL, "");

+#ifdef _WIN32
+    SetConsoleOutputCP(65001);
+#endif
+
     struct criterion_test test;
     memset(&test, 0, sizeof (test));
     struct criterion_suite suite;
diff --git a/src/entry/params.c b/src/entry/params.c
index 04fd47b..0a162cd 100644
--- a/src/entry/params.c
+++ b/src/entry/params.c
@@ -279,7 +279,6 @@ CR_API int criterion_handle_args(int argc, char *argv[],
         { 0,                 0,                 0, 0   }
     };

-    setlocale(LC_ALL, "");
 #if ENABLE_NLS
     textdomain(PACKAGE "-test");
 #endif
```

**Comment by Snaipe** [comment]
This issue is somewhat vexing. I'm currently reviewing my old branch, as well as the comments above, and the way windows implements the printf function family doesn't make any god damn sense.

After all the dancing around with vsnprintf to try to make things right, I would somewhat expect to have consistent results; but as it turns out, they're not. `vsnprintf(buf, sz, "%ls", L"ą")` works as expected and produces an UTF-8 `"ą"`, which gets correctly printed by `printf` later on, but for some reason `vsnprintf(buf, zs, "%s", "ą")` produces "a".

I'm amazed by the level at which the printf family is broken on Windows; so much that I'm really tring hard not to be tempted to just throw it all away and use a *good* implementation rather than trying to make Windows' standard C library work.

Rant aside, I think there's something going on with the locale names given to setlocale that need to be windows-specific; I'll have a more in-depth look at it.

**Comment by am11** [comment]
Could we update https://github.com/Snaipe/Criterion/blob/ccf5d1dcd66c81d32536b043a4df608a72c7f6ca/appveyor.yml#L3 to `Visual Studio 2017` to check if it makes any difference in vsnprintf behavior? Seems like they have made some improvements around C99/C11 to make backend compiler compatible with LLVM/Clang6.0 frontend [blog](https://blogs.msdn.microsoft.com/vcblog/2018/05/07/announcing-msvc-conforms-to-the-c-standard/).

**Comment by Snaipe** [comment]
I think I'll had a matrix entry to test on Visual Studio 2017, since I remember that VS2017 broke things that previously worked on VS2015, but the main issue remains on MinGW due to locale awareness & abysmal locale support on Windows messing things up.

I've had a look at https://github.com/mpaland/printf and I think this might be a viable alternative to integrate on Windows systems; though there are a few things left to be 100% compatible with our use case, like supporting all the standard length modifiers.
